openadapt/crud.py

```diff
@@ -221,19 +214,37 @@ def filter_stop_sequences(action_events):
         for _ in range(0, num_to_remove):
             action_events.pop()
 
+def save_screenshot_diff(screenshots):
+    data_updated = False
+    logger.info("verifying diffs for screenshots...")
+
+    for screenshot in screenshots:
+        if not screenshot.prev:
+            continue
+        if not screenshot.png_diff_data:
+            screenshot.png_diff_data = screenshot.convert_png_to_binary(screenshot.diff)
+            data_updated = True
+        if not screenshot.png_diff_mask_data:
+            screenshot.png_diff_mask_data = screenshot.convert_png_to_binary(screenshot.diff_mask)
+            data_updated = True
+
+    if data_updated:
+        logger.info("saving screenshot diff data to db...")
+        db.bulk_save_objects(screenshots)
+        db.commit()
+
+    return screenshots
+
 
-def get_screenshots(recording, precompute_diffs=False):
+def get_screenshots(recording):
     screenshots = _get(Screenshot, recording.timestamp)
 
     for prev, cur in zip(screenshots, screenshots[1:]):
         cur.prev = prev
     screenshots[0].prev = screenshots[0]
 
-    # TODO: store diffs
-    if precompute_diffs:
-        logger.info(f"precomputing diffs...")
-        [(screenshot.diff, screenshot.diff_mask) for screenshot in screenshots]
-
+    if config.SAVE_SCREENSHOT_DIFF:
```

Comment:
@LaPetiteSouris What do you think about automatically precomputing diffs if they are not already pre-computed?

Reply:
Doing this would ensure that we have all necessary data when we need advanced training. However, as I mentioned, it would incur heavy computing/CPU resources on first time running.

Ideally, I think we can provide 2 tracks:
- One is "basic/fast" track, which allow user to bypass all of this and have a naive but really fast results.
- One "advanced" track which incur heavy computation but provides best results.

This approach, from product side is well-round IMHO. However, as we have not figured out what and how to provide 2 experiences, I still suggest to make this computation opt-in.

WDYT ?

Reply:
Sounds good!

Reply:
Thanks. Is there anything else you would request on this PR before an `LGTM` ? 🥇